### PR TITLE
Add UpSampling2D Layer

### PR DESCRIPTION
**Pull request by tanmayb123**
Add UpSampling2D layer. Implementation based off of: https://github.com/tensorflow/tpu/blob/c9d7aa72464751c26e33a3db4985129b4788baa1/models/official/retinanet/retinanet_architecture.py#L420

### CODE REVIEW

File: Sources/DeepLearning/Layer.swift [hunk 1]
```diff
@@ -367,3 +367,36 @@ public struct Dropout<Scalar>: Layer
...
     }
 }
+
+@_fixed_layout
+public struct UpSampling2D<Scalar>: Layer
+    where Scalar: BinaryFloatingPoint & Differentiable & TensorFlowScalar {
+
+    @noDerivative public let scale: Int32
+
+    public init(scaleBy: Int32) {
```
Comment (rxwei) [inline Sources/DeepLearning/Layer.swift]: In initializers, argument labels and arguments do not need to form a phrase. Use `scalar:` instead.

File: Sources/DeepLearning/Layer.swift [hunk 2]
```diff
@@ -367,3 +367,36 @@ public struct Dropout<Scalar>: Layer
...
+    }
+
+    @differentiable(wrt: (self, input))
+    public func applied(to input: Tensor<Scalar>) -> Tensor<Scalar> {
+        let batchSize = input.shape[0]
+        let height = input.shape[1]
+        let width = input.shape[2]
+        let channels = input.shape[3]
+        let upSampling = input.reshaped(
+          toShape: Tensor<Int32>([batchSize,
```
Comment (rxwei) [inline Sources/DeepLearning/Layer.swift]: Break this down to multiple statements: `let shape =`, `Tensor<Int32>([...])`, and `Tensor(ones:)`.

**Review by rxwei — COMMENTED** [review]
Thanks!

File: Sources/DeepLearning/Layer.swift (hunk 1 above)
Comment (tanmayb123) [reply Sources/DeepLearning/Layer.swift]: Alright. Do you think `size:` would fit better than `scalar:`? (I'm basing this off of https://github.com/keras-team/keras/blob/master/keras/layers/convolutional.py#L1944)

Comment (rxwei) [reply Sources/DeepLearning/Layer.swift]: Just noticed I had a typo: `scale` not `scalar` :) 

Yes, `size` sounds better.

File: Sources/DeepLearning/Layer.swift (hunk 2 above)
Comment (tanmayb123) [reply Sources/DeepLearning/Layer.swift]: Fixed.

File: Sources/DeepLearning/Layer.swift (hunk 1 above)
Comment (tanmayb123) [reply Sources/DeepLearning/Layer.swift]: Fixed.

File: Sources/DeepLearning/Layer.swift [hunk 3]
```diff
@@ -367,3 +367,36 @@ public struct Dropout<Scalar>: Layer
...
+        self.size = size
+    }
+
+    @differentiable(wrt: (self, input))
+    public func applied(to input: Tensor<Scalar>) -> Tensor<Scalar> {
+        let batchSize = input.shape[0]
+        let height = input.shape[1]
+        let width = input.shape[2]
+        let channels = input.shape[3]
+        let reshapeSize = Tensor<Int32>([batchSize,
```
Comment (rxwei) [inline Sources/DeepLearning/Layer.swift]: One minor issue: You don't have to break lines for any of these statements, as long as everything fits within 100 columns.

**Review by rxwei — APPROVED** [review]

File: Sources/DeepLearning/Layer.swift (hunk 3 above)
Comment (tanmayb123) [reply Sources/DeepLearning/Layer.swift]: Done.

Please let me know if there're any more issues I need to fix.

Comment (rxwei) [reply Sources/DeepLearning/Layer.swift]: Sorry, i wasn't super clear -- i was referring to the lines that are broken for the array literal, not the `let`s.

I actually liked temporary variables for dimension names :)

Comment (rxwei) [reply Sources/DeepLearning/Layer.swift]: Also, you can bring the right multiplication operand `scaleOnes` to its previous line.

Comment (tanmayb123) [reply Sources/DeepLearning/Layer.swift]: Ah, alright, thanks. Fixed.

**Comment by rxwei** [comment]
Looks great. Thanks Tanmay!
